Port Main.DSP to OnionConfig

malleoz opened the pull request:
While trying to work on adding audiodump support for CLI, I was alerted that it was important to first try moving the DSP configs to the new OnionConfig before continuing, as that makes it substantially easier to write clean code to add such a feature.

This PR aims to allow for Dolphin to only rely on OnionConfig for DSP-related settings.

I am not very familiar with the new config, so please let me know if I missed any concepts in this PR. Notably, I was unsure what to do in BootManager.cpp `BootCore()` in the case of netplay... Seems like it uses its own SConfig which checks for DSP JIT.

JosJuice commented on Source/Core/Core/ConfigLoaders/IsSettingSaveable.cpp:
```diff
@@ -72,6 +72,20 @@ bool IsSettingSaveable(const Config::Location& config_location)
+      &Config::MAIN_AUDIO_MUTED.GetLocation(),
+#ifdef _WIN32
+      &Config::MAIN_WASAPI_DEVICE.GetLocation(),
+#endif
```
Instead of adding all of these, you could add "DSP" to the list of sections at line 29.

Also, I think you've forgotten MAIN_DSP_HLE.

JosJuice replied:
Actually, speaking of DSP HLE, you don't seem to have removed that one from ConfigManager.cpp. Please be consistent with whether you're intending to port that one over or not. If it's not going to be in the new config system, then you shouldn't be adding calls to `Config::Get(Config::MAIN_DSP_HLE)`.

malleoz replied:
Added DSP to line 29, thanks for that cleanup. I had forgotten MAIN_DSP_HLE, so I've added that in `s_setting_saveable` under Main.Core.

You're right, I hadn't removed it from ConfigManager.cpp but I've done so now. There was an SConfig function dependent on bDSPHLE and bDPL2Decoder, so I ported both over now.

JosJuice commented on Source/Core/Core/ConfigLoaders/NetPlayConfigLoader.cpp:
```diff
@@ -47,6 +47,7 @@ class NetPlayConfigLayerLoader final : public Config::ConfigLayerLoader
     layer->Set(Config::MAIN_FALLBACK_REGION, m_settings.m_FallbackRegion);
     layer->Set(Config::MAIN_ALLOW_SD_WRITES, m_settings.m_AllowSDWrites);
     layer->Set(Config::MAIN_DSP_JIT, m_settings.m_DSPEnableJIT);
+    layer->Set(Config::MAIN_DSP_HLE, m_settings.m_DSPHLE);
```
There's already a line that sets this.

JosJuice commented:
@dolphin-emu-bot rebuild

JosJuice commented:
Also, the new config system is not officially called OnionConfig. But this is just me being opinionated and not something that necessarily should block the PR :)

malleoz commented on Source/Core/Core/Config/MainSettings.cpp:
```diff
@@ -129,13 +128,18 @@ const Info<bool> MAIN_DISABLE_SCREENSAVER{{System::Main, "Display", "DisableScre
 // Main.DSP
 
 const Info<bool> MAIN_DSP_CAPTURE_LOG{{System::Main, "DSP", "CaptureLog"}, false};
+const Info<bool> MAIN_DSP_HLE{{System::Main, "DSP", "DSPHLE"}, true};
```
It was originally in core prior to this commit, but I thought, given that fact that MAIN_DSP_JIT was under DSP category, that it was fitting to move HLE to this section too. I might be misunderstanding something though.

JosJuice replied:
Changing which INI section a setting is stored in breaks existing configs, which should not be done unless there's some particularly important reason. Had you been adding a new setting then I agree that putting it in DSP makes sense, but this little mismatch isn't important enough to warrant breaking configs.

malleoz replied:
Ah I see. I will move it back to Core, where it was originally.

malleoz commented on Source/Core/Core/Config/MainSettings.h:
```diff
@@ -9,6 +9,8 @@
 #include "Common/Config/Config.h"
 #include "DiscIO/Enums.h"
 
+#define BACKEND_NULLSOUND _trans("No Audio Output")
```
It's referenced in AudioCommon.cpp which was originally using it from the old config file... But I now realize I should've moved all the others over... Not sure why I didn't.

JosJuice replied:
Ah, I didn't think of the fact that it was in ConfigManager.h originally. Yes, in that case I think it would make sense to move them all here.

malleoz commented:
Implemented Jos's review notes. Kept `MAIN_DSP_HLE` in Core so as to not break previous config files, finished porting over DSP HLE to use the new config, removed references to the new config from `ConfigCache`, and ported over `MAIN_DSP_THREAD` too.

JosJuice commented on Source/Core/AudioCommon/AudioCommon.cpp:
```diff
@@ -241,7 +242,7 @@ void DecreaseVolume(unsigned short offset)
 void ToggleMuteVolume()
 {
-  bool& isMuted = SConfig::GetInstance().m_IsMuted;
+  bool isMuted = Config::Get(Config::MAIN_AUDIO_MUTED);
```
This doesn't do the same thing anymore. (Note how the reference is assigned to in the next line.) Same with `currentVolume` in two places above.

malleoz replied:
Ahh, I understand why a reference was used here now. Will fix this.
